pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -0,0 +1,129 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gkenetworkparamset
+
+import (
+	"golang.org/x/net/context"
+	"google.golang.org/api/compute/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	networkv1alpha1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1alpha1"
+	networkclientset "k8s.io/cloud-provider-gcp/crd/client/network/clientset/versioned"
+	networkinformers "k8s.io/cloud-provider-gcp/crd/client/network/informers/externalversions/network/v1alpha1"
+	"k8s.io/cloud-provider-gcp/pkg/util/gkenetworkparamsetutil"
+	"k8s.io/cloud-provider-gcp/providers/gce"
+	controllersmetrics "k8s.io/component-base/metrics/prometheus/controllers"
+)
+
+// Controller manages GKENetworkParamSet status.
+type Controller struct {
+	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer
+	networkClientset         networkclientset.Interface
+	informerStartFn          func(stopCh <-chan struct{})
+	gceCloud                 *gce.Cloud
+}
+
+// NewGKENetworkParamSetController returns a new
+func NewGKENetworkParamSetController(
+	networkClientset networkclientset.Interface,
+	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer,
+	informerStartFn func(stopCh <-chan struct{}),
+	gceCloud *gce.Cloud,
+) *Controller {
+
+	return &Controller{
+		networkClientset:         networkClientset,
+		gkeNetworkParamsInformer: gkeNetworkParamsInformer,
+		informerStartFn:          informerStartFn,
+		gceCloud:                 gceCloud,
+	}
+
+}
+
+// Run starts an asynchronous loop that monitors and updates GKENetworkParamSet in the cluster.
+func (c *Controller) Run(stopCh <-chan struct{}, controllerManagerMetrics *controllersmetrics.ControllerManagerMetrics) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting gkenetworkparamset controller")
+	defer klog.Infof("Shutting down gkenetworkparamset controller")
+	controllerManagerMetrics.ControllerStarted("gkenetworkparamset")
+	defer controllerManagerMetrics.ControllerStopped("gkenetworkparamset")
+
+	c.gkeNetworkParamsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    gkenetworkparamsetutil.CreateAddGKENetworkParamSetHandler(c.onGKENetworkParamSetAdded),
+		UpdateFunc: gkenetworkparamsetutil.CreateUpdateGKENetworkParamSetHandler(c.onGKENetworkParamSetUpdated),
+	})
+
+	c.informerStartFn(stopCh)
+
+	<-stopCh
+	klog.Infof("done gkenetworkparamset controller")
+}
+
+func (c *Controller) onGKENetworkParamSetAdded(params *networkv1alpha1.GKENetworkParamSet) error {
+	return c.updateGKENetworkParamSet(params)
+}
+func (c *Controller) onGKENetworkParamSetUpdated(oldParams, newParams *networkv1alpha1.GKENetworkParamSet) error {
+	return c.updateGKENetworkParamSet(newParams)
+}
+
+func (c *Controller) updateGKENetworkParamSet(params *networkv1alpha1.GKENetworkParamSet) error {
```

Comment:
Originally, retry logic was handled by the resync feature of the informer that caused this function to be rerun periodically. We didn't want a periodic refetch of the subnetwork details after the object status had been resolved so that it wouldn't update the resource if the subnetwork is modified after resolution in GCP.

After further discussion, it seemed like we would want some more advanced retry logic beyond the informer resync, along with event logging on the resource.

My judgment was to put up this PR which covers the happy path (so we can test) and put in the error/retry logic next.

Reply:
Sounds good.